backup: include tenant usage data in backup

RaduBerinde opened the pull request:
#### backup: minor refactoring of tenant backup planning

This commit extracts the code that generates the tenant metadata into
a separate file.

Release note: None

Release justification: Part of a larger change that is necessary to
fully support backup and restore of non-system tenants. The change
only concerns multi-tenant deployments and should not affect single
tenant deployments.

#### backup: include tenant usage data in backup

We now back up and restore relevant data from tenant_usage: the last
token bucket state, and (more importantly) total all-time tenant
consumption.

The backup manifest proto had to change, but we still support
restoring tenant data from older manifests.

Release note: None

Release justification: This change is necessary to fully support
backup and restore of non-system tenants. The change only concerns
multi-tenant deployments and should not affect single tenant
deployments.

Informs #68479.

#### tenantcostserver: check tenant when initializing state

The tenant_usage data is initialized as needed unless it is configured
explicitly. This commit adds validation of the tenant in this case,
preventing the state from appearing for a deleted tenant.

Release note: None

Release justification: Part of a larger change that is necessary to
fully support backup and restore of non-system tenants. The change
only concerns multi-tenant deployments and should not affect single
tenant deployments.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/70595)
<!-- Reviewable:end -->

dt commented on pkg/jobs/jobspb/jobs.proto:
```diff
@@ -195,7 +195,8 @@ message RestoreDetails {
   repeated sqlbase.SchemaDescriptor schema_descs = 15;
-  repeated sqlbase.TenantInfo tenants = 13 [(gogoproto.nullable) = false];
+  reserved 13;
+  repeated sqlbase.TenantInfoWithUsage tenants = 22 [(gogoproto.nullable) = false];
```
was this supposed to be 21?

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70595#-:-MkO2AeM5wsFI1T0YNCK:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, @dt, @nvanbenschoten, and @tbg)

___
*[pkg/jobs/jobspb/jobs.proto, line 199 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/70595#-MkEMi5imCFT-N9INFL--r2-199:-MkFUGMB2i7Bkpg4oLvU:b2nxmgb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ddb188196aaf8757e6a250caa5d7dd6f2ea0ebb7/pkg/jobs/jobspb/jobs.proto#L199)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

was this supposed to be 21?
</blockquote></details>

Indeed, good catch!


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70595)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, @dt, @nvanbenschoten, and @tbg)

___
*[pkg/sql/tenant.go, line 95 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70595#-MkSDahU4rNNnzrIN2Ug:-MkSDahU4rNNnzrIN2Uh:b9a269w) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ada78d445baf57e512e95f683fa40b1b07f594d/pkg/sql/tenant.go#L95)):*
> ```Go
> 		return errors.Wrap(err, "inserting new tenant")
> 	} else if num != 1 {
> 		log.Fatalf(ctx, "unexpected number of rows affected: %d", num)
> ```

I'm investigating a test failure under stressrace. While trying to repro, I hit this. I'll try to figure out what's wrong with the test.

But should this be a Fatal? Seems like if you somehow try to destroy a tenant two times in parallel, you could hit this; doesn't seem right to crash the node..


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Added a commit - we now check the tenant ID when we initialize the tenant usage state. This eliminates a race in the test, where existing tenant processes could cause re-creation of the usage state after the tenant was deleted.

RaduBerinde committed:
tenantcostserver: check tenant when initializing state

The tenant_usage data is initialized as needed unless it is configured
explicitly. This commit adds validation of the tenant in this case,
preventing the state from appearing for a deleted tenant.

Release note: None

Release justification: Part of a larger change that is necessary to
fully support backup and restore of non-system tenants. The change
only concerns multi-tenant deployments and should not affect single
tenant deployments.

RaduBerinde committed:
backup: include tenant usage data in backup

We now back up and restore relevant data from tenant_usage: the last
token bucket state, and (more importantly) total all-time tenant
consumption.

The backup manifest proto had to change, but we still support
restoring tenant data from older manifests.

Release note: None

Release justification: This change is necessary to fully support
backup and restore of non-system tenants. The change only concerns
multi-tenant deployments and should not affect single tenant
deployments.

RaduBerinde committed:
backup: minor refactoring of tenant backup planning

This commit extracts the code that generates the tenant metadata into
a separate file.

Release note: None

Release justification: Part of a larger change that is necessary to
fully support backup and restore of non-system tenants. The change
only concerns multi-tenant deployments and should not affect single
tenant deployments.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70595)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @andy-kimball, @dt, @nvanbenschoten, and @tbg)

___
*[pkg/sql/tenant.go, line 95 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/70595#-MkSDahU4rNNnzrIN2Ug:-MkZNP944hqiGanG64ps:b-o9chjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ada78d445baf57e512e95f683fa40b1b07f594d/pkg/sql/tenant.go#L95)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I'm investigating a test failure under stressrace. While trying to repro, I hit this. I'll try to figure out what's wrong with the test.

But should this be a Fatal? Seems like if you somehow try to destroy a tenant two times in parallel, you could hit this; doesn't seem right to crash the node..
</blockquote></details>

Sorry, I pointed to the wrong code, the question was about the Fatalf after the DELETE TENANT query in GCTenantSync below


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
